### PR TITLE
Add support for merging metadata from formatting params

### PR DESCRIPTION
**Pull request by danielchatfield**
Within our codebase it is common to slog an error using code of the following form:

```
slog.Error(ctx, "Error performing foo: %v", err, slogParams)
```

Often the error contains useful debugging information in its terror params but these don't make it into the slog event. This encourages stuffing that data into the terror message which makes it more challenging to work with programatically.

This change introduces a `logMetadataProvider` interface, any param that satisfies this interface will automatically have the metadata merged into the slog event metadata.

### CODE REVIEW

File: event.go [hunk 1]
```diff
@@ -11,13 +11,13 @@ import (
...
-	WarnSeverity
-	ErrorSeverity
-	CriticalSeverity
+	TimeFormat                = "2006-01-02 15:04:05-0700 (MST)"
+	TraceSeverity    Severity = 1
+	DebugSeverity    Severity = 2
+	InfoSeverity     Severity = 3
+	WarnSeverity     Severity = 4
+	ErrorSeverity    Severity = 5
+	CriticalSeverity Severity = 6
```
Comment (danielchatfield) [inline event.go]: The behaviour of `iota` is a little non intuitive so I think being explicit here is better, this is unrelated to the other changes.

File: .travis.yml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 language: go
 
 go:
-  - 1.6.3
-  - 1.7
+  - 1.8.x
+  - 1.9.x
```
Comment (danielchatfield) [inline .travis.yml]: travis was failing without this change

Comment (obeattie) [reply .travis.yml]: Could you also add Go 1.10 as a target while you're here?
